TransformationSystem/DB/TransformationDB.py

```diff
@@ -1417,52 +1415,50 @@ def exists( self, lfns, connection = False ):
-         In the input dict, the only mandatory info are PFN and SE
+    """ Add the supplied lfn to the Transformations and to the DataFiles table if it passes the filter
     """
     gLogger.info( "TransformationDB.addFile: Attempting to add %s files." % len( fileDicts ) )
     successful = {}
     failed = {}
     # Determine which files pass the filters and are to be added to transformations
     transFiles = {}
     filesToAdd = []
+    catalog = FileCatalog()
+
     for lfn in fileDicts:
-      fileTrans = self.__filterFile( lfn )
-      if not ( fileTrans or force ):
-        successful[lfn] = True
+      gLogger.info( "addFile: Attempting to add file %s" % lfn )
+      res = catalog.getFileUserMetadata( lfn )
+      if not res['OK']:
+        failed[lfn] = res['Message']
+        return S_OK( {'Successful':successful, 'Failed':failed } )
+      else:
+        metadatadict = res['Value']
+      gLogger.info( 'Filter file with metadata', metadatadict )
+      fileTrans = self._filterFileByMetadata( metadatadict )
+      gLogger.info('fileTrans', fileTrans)
+      if not ( fileTrans or force ):  # not clear how force should be used for
+        successful[lfn] = False  # True -> False bug fix: otherwise it is set to True even if fileTrans is empty.
       else:
         filesToAdd.append( lfn )
         for trans in fileTrans:
-          transFiles.setdefault( trans, [] ).append( lfn )
-    # Add the files to the DataFiles and Replicas tables
-    if filesToAdd:
-      connection = self.__getConnection( connection )
-      res = self.__addDataFiles( filesToAdd, connection = connection )
-      if not res['OK']:
-        return res
-      lfnFileIDs = res['Value']
-      for lfn in filesToAdd:
-        if lfn in lfnFileIDs:
-          successful[lfn] = True
-        else:
-          failed[lfn] = True
+          if not transFiles.has_key( trans ):
+            transFiles[trans] = []
+          transFiles[trans].append( lfn )
+
       # Add the files to the transformations
-      # TODO: THIS SHOULD BE TESTED WITH A TRANSFORMATION WITH A FILTER
-      for transID, lfns in transFiles.items():
-        fileIDs = []
-        for lfn in lfns:
-          if lfn in lfnFileIDs:
-            fileIDs.append( lfnFileIDs[lfn] )
-        if fileIDs:
-          res = self.__addFilesToTransformation( transID, fileIDs, connection = connection )
+      gLogger.info( 'Files to add to transformations:', filesToAdd )
+      if filesToAdd:
+        for transID, lfns in transFiles.items():
+          res = self.addFilesToTransformation( transID, lfns )
           if not res['OK']:
-            gLogger.error( "Failed to add files to transformation", "%s %s" % ( transID, res['Message'] ) )
-            failed[lfn] = True
-            successful[lfn] = False
+            for lfn in lfns:
+              gLogger.error( "Failed to add files to transformation", "%s %s" % ( transID, res['Message'] ) )
+              failed[lfn] = res['Message']
```

Comment:
Yes. Do you suggest for instance replacing with:

if not res['OK']:
  gLogger.error( "Failed to add files to transformation", "%s %s" % ( transID, res['Message'] ) )
  return res
 else:
   successful.update( res['Value']['Successful'] )
    failed.update( res['Value']['Failed'] )

?